### PR TITLE
raft: send promote only after new term has been broadcasted

### LINKED ISSUE #6754: PROMOTE for the next term sometimes arrives earlier than the term bump
Try running the following test:
<details><summary>bug_test.lua</summary>

```lua
local luatest = require('luatest')
local helpers = require('test.luatest_helpers')
local cluster = require('test.luatest_helpers.cluster')
local g = luatest.group('bug')

g.before_each(function(g)
    g.cluster = cluster:new({})

    g.box_cfg = {
        election_mode = 'off';
        replication = {
            helpers.instance_uri('server_', 1);
            helpers.instance_uri('server_', 2);
        };
    }

    g.server_1 = g.cluster:build_and_add_server(
        {alias = 'server_1', box_cfg = g.box_cfg})
    g.server_2 = g.cluster:build_and_add_server(
        {alias = 'server_2', box_cfg = g.box_cfg})
    g.cluster:start()
end)

g.after_each(function(g)
    g.cluster:stop()
    g.cluster.servers = nil
end)

g.test_interfering_promote_in_wal = function(g)
    local wal_write_count = g.server_1:exec(function()
        box.error.injection.set('ERRINJ_WAL_DELAY', true)
        return box.error.injection.get('ERRINJ_WAL_WRITE_COUNT')
    end)

    g.server_2:exec(function()
        box.ctl.promote()
    end)
    g.server_1:wait_wal_write_count(wal_write_count + 2)

    local ok, err = g.server_1:exec(function()
        local f = require('fiber').new(box.ctl.promote); f:set_joinable(true)
        box.error.injection.set('ERRINJ_WAL_DELAY', false)
        return f:join()
    end)

    luatest.assert(
        not ok and err.code == box.error.INTERFERING_PROMOTE,
        'interfering promote not handled')
end
```

</details>

`./test-run.py $(yes replication-luatest/bug_test.lua | head -n 1000) -j16`

Sooner or later all the workers will hang.

The workers hang for the following reason:
Note, after `server_2` has issues `box.ctl.promote()`, we wait for 2 wal write attempts on `server_1`:
```lua
g.server_1:wait_wal_write_count(wal_write_count + 2)
```
This is because we assume, that `server_1` will first receive the term bump (`IPROTO_RAFT`), and only then the promote request itself.

When the waiting for 2 wal writes times out, this is because PROMOTE arrives first, and applier is blocked on a synchronous WAL write, so it never reads the term bump and never attempts the second write.

A PROMOTE for a specific term arriving on a replica earlier than the term bump itself might be bad: we assume everywhere that txn_limbo's term (bumped by promote) always follows raft term.

In order to fix this we might either write server's full raft state together with a PROMOTE or, maybe, relay raft term updates from WAL instead of relying on asynchronous notifications.

Found by @grafin while working on #6033

### PR DESCRIPTION
**Pull request by grafin**
Wait until latest term will be sent out to all active replicas before writing promote.

Closes #6754

### CODE REVIEW

File: src/box/relay.cc
Comment (sergepetrenko) [inline src/box/relay.cc:1103]: Why don't you run the triggers when there's a new broadcast? It might be not related to term anyhow. Even if it is related to term (e.g. another term bump), the logic in `box_raft_wait_term_broadcasted` will compare `sent_term` with `volatile_term` and everything will work correctly.

The trigger should run on every successful message despatch.

File: src/box/raft.c
Comment (grafin) [reply src/box/raft.c:481]: box_raft_wait_term_broadcasted_f will be called only if term was sent to someone.
yield will fail if broadcast was scheduled in replicaset of two replicas when relay was alive, first check here yeilds, then relay dies for some reason before term is sent-out by relay, we will never return here.

Comment (grafin) [reply src/box/raft.c]: UPD. I've tried changing it to fiber_yield() and it passes all tests on my machine:) I changed it to yield. But still have some doubts.

Comment (sergepetrenko) [reply src/box/raft.c]: >  then relay dies for some reason before term is sent-out by relay, we will never return here.

I agree,  that's possible. Our tests might miss that schenario.
I suggest you call box_raft_on_message_send() when relay exits due to any reason.
This way we won't have to spin in a busy loop here and won't lose any message dispatch events.

File: src/box/box.cc
Comment (sergepetrenko) [reply src/box/box.cc]: As I see we don't check interfering promotes in `box_promote_qsync`. Let's leave it as is then.

File: src/box/box.cc
Comment (sergepetrenko) [inline src/box/box.cc:1663]: Consider this diff:
```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 1687a003d..26659510a 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1660,9 +1660,6 @@ box_run_elections(void)
        struct raft *raft = box_raft();
        assert(raft->is_enabled);
        assert(box_election_mode != ELECTION_MODE_VOTER);
-       if (box_raft_wait_term_broadcasted() != 0)
-               return -1;
-
        /*
         * Make this instance a candidate and run until some leader, not
         * necessarily this instance, emerges.
@@ -1674,6 +1671,10 @@ box_run_elections(void)
                        return -1;
                }
        } while (raft->leader == 0);
+
+       if (box_raft_wait_term_broadcasted() != 0)
+               return -1;
+
        if (raft->state != RAFT_STATE_LEADER) {
                diag_set(ClientError, ER_INTERFERING_PROMOTE,
                         raft->leader);

```

File: src/box/box.cc
Comment (sergepetrenko) [inline src/box/box.cc:1939]: Extraneous change. Please, remove.

File: src/box/relay.cc
Comment (sergepetrenko) [reply src/box/relay.cc:1103]: Never mind. I misread this code chunk

**Review by sergepetrenko — CHANGES_REQUESTED** [review]
Thanks! 2 minor comments remain, everything else looks fine.

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:1844]: Here `raft->term == raft->volatile_term`, because the state is `RAFT_STATE_LEADER` as checked a few lines above. It means the term is already on disk and is already pushed to the relay threads. Doesn't it mean that you can write promote without any waiting and it won't overtake the term update anyway?

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:1846]: If the problem is only with promote/demote, can this waiting be done right inside of `box_issue_promote/demote()`?

File: src/box/raft.c
Comment (Gerold103) [inline src/box/raft.c:58]: It seems you don't need triggers. You would be fine with a simple `fiber_cond`. Because the only user of this trigger is a blocking wait. `box_raft_wait_term_broadcasted` would wait on the cond. This cond could be called `box_raft_on_message_cond`.

File: src/box/raft.c
Comment (Gerold103) [inline src/box/raft.c:451]: Lets fix the indentation. Multiline `if`s need to be aligned by the opening brace + 1:
```C
if (replica->relay == NULL || replica->anon ||
    tt_uuid_is_equal(&replica->uuid, &INSTANCE_UUID))
```
The same below.

File: src/box/raft.c
Comment (Gerold103) [inline src/box/raft.c:447]: `recieve` -> `receive`.

File: src/box/raft.c
Comment (Gerold103) [inline src/box/raft.c:483]: The problem here is that your function will wait until `term == volatile_term`. If the volatile term is bumped during the wait, you will continue the waiting. Even though it becomes unnecessary, because the promotion has already failed at this point. Also it might make the waiting potentially quite long if term bumps will start happening very often.
I suggest to remember the volatile term you are waiting for and wait for any persistent term >= the remembered one. 

Then the caller will need to check if the term has changed during the waiting. And most likely abort whatever it was doing.

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the patch! That is quite a bug you've found, nice!

File: src/box/box.cc
Comment (grafin) [reply src/box/box.cc:1844]: Thanks for indepth review! The problem is term might be in relay for some time, during that time promote can be written to WAL and broadcasted by relay before term is. AFAIK there is no guarantee that relay will broadcast new term, before it broadcasts promote from wal.

File: src/box/box.cc
Comment (grafin) [reply src/box/box.cc:1846]: It can, the only problem is there is no return value from box_issue_promote/demote. Should I move it there anyway?

File: src/box/box.cc
Comment (Gerold103) [reply src/box/box.cc:1844]: Ok, I see, then lets leave it.

File: src/box/box.cc
Comment (Gerold103) [reply src/box/box.cc:1846]: Your function can only fail if the fiber is cancelled. But `box_issue_promote/demote()` are writing to disk. They are non-cancellable during that procedure. It means that you could make `box_raft_wait_term_broadcasted()` never failing and wait regardless of whether the fiber is cancelled. And it won't matter in case you will do it right before the WAL write. I don't think it would cause any problems. Lets try?

**Comment by Gerold103** [comment]
There is an alternative and potentially simpler solution. I discussed it with Sergey and we think it is worth a try.

The problem we are trying to solve here is that the relay sees term bumps in WAL, but it ignores them because they belong to `GROUP_LOCAL`. On the other hand, we can't use these bumps from WAL, because we also need at least vclock and will need to know who is the leader (for the future fencing).

The proposal is lets in the relay thread keep track of which terms were already sent. Like you do now, but only in the relay thread, not in tx. Then lets not ignore term bumps from WAL. Instead, we compare them to already sent terms. If the term from WAL is > last sent, it means TX thread is trying to send us a term update, but we didn't receive it yet. Then we spin in `cbus_process()` waiting for that cmsg from tx thread. When received - continue.

The best part of that solution is that it is located entirely in the relay thread. TX is not changed anyhow. Only relay.cc and relay.h would change, ideally. And only their parts working in the relay thread.

Could you please implement that on another branch so we could compare which option looks better in the end? Feel free to ask questions.

**Comment by grafin** [comment]
@Gerold103 Thanks for review and improvement suggestions. I've tried checking for sent term in relay, the problem is cbus_process moves messages to local variable (in cbus_endpoint_fetch) if in that moment both promote and term bump messages are already in the endpoint in wrong order (promote before term bump) nothing can be done outside cbus to restore the order.
Because of that I've reworked this PR using fiber_cond, and for me it looks like the best solution right now.

I've tried creating a stable reproducer but failed to do so. If you have any ideas on how to do that, please let me know.

**Comment by Gerold103** [comment]
> if in that moment both promote and term bump messages are already in the endpoint in wrong order (promote before term bump) nothing can be done outside cbus to restore the order.

How is it possible? Promote is not sent via cbus at all. Relay reads it from WAL. In cbus you only get a term update. So inside of cbus there can't be a race. You now have only a race of WAL vs `relay<->tx` cbus. When you read promote from WAL before receiving a new term from cbus. Could you please show me the patch? I will try to figure out what went wrong.

File: changelogs/unreleased/gh-6754-term-bump-before-promote.md
Comment (Gerold103) [inline changelogs/unreleased/gh-6754-term-bump-before-promote.md:1]: I managed to create a stable reproducer, see diff below and a test:
<details><summary>diff</summary>

```Diff
diff --git a/src/box/lua/info.c b/src/box/lua/info.c
index 8d7bb1d4a..76cc50ada 100644
--- a/src/box/lua/info.c
+++ b/src/box/lua/info.c
@@ -644,6 +644,8 @@ lbox_info_synchro(struct lua_State *L)
 	lua_setfield(L, -2, "owner");
 	lua_pushboolean(L, latch_is_locked(&queue->promote_latch));
 	lua_setfield(L, -2, "busy");
+	luaL_pushuint64(L, queue->promote_greatest_term);
+	lua_setfield(L, -2, "term");
 	lua_setfield(L, -2, "queue");
 
 	return 1;
diff --git a/src/box/relay.cc b/src/box/relay.cc
index 1c588f70a..8281b7656 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -141,8 +141,14 @@ struct relay {
 	 */
 	struct vclock local_vclock_at_subscribe;
 
-	/** Relay endpoint */
-	struct cbus_endpoint endpoint;
+	/** Endpoint to receive messages from WAL. */
+	struct cbus_endpoint wal_endpoint;
+	/**
+	 * Endpoint to receive messages from TX. Having the 2 endpoints
+	 * separated helps to synchronize the data coming from TX and WAL. Such
+	 * as term bumps from TX with PROMOTE rows from WAL.
+	 */
+	struct cbus_endpoint tx_endpoint;
 	/** A pipe from 'relay' thread to 'tx' */
 	struct cpipe tx_pipe;
 	/** A pipe from 'tx' thread to 'relay' */
@@ -811,7 +817,7 @@ relay_send_is_raft_enabled(struct relay *relay,
 	 * fiber.
 	 */
 	while (!msg->is_finished) {
-		cbus_process(&relay->endpoint);
+		cbus_process(&relay->tx_endpoint);
 		if (msg->is_finished)
 			break;
 		fiber_yield();
@@ -831,12 +837,16 @@ relay_subscribe_f(va_list ap)
 	coio_enable();
 	relay_set_cord_name(relay->io->fd);
 
-	/* Create cpipe to tx for propagating vclock. */
-	cbus_endpoint_create(&relay->endpoint, tt_sprintf("relay_%p", relay),
+	cbus_endpoint_create(&relay->tx_endpoint,
+			     tt_sprintf("relay_tx_%p", relay),
 			     fiber_schedule_cb, fiber());
-	cbus_pair("tx", relay->endpoint.name, &relay->tx_pipe,
+	cbus_pair("tx", relay->tx_endpoint.name, &relay->tx_pipe,
 		  &relay->relay_pipe, NULL, NULL, cbus_process);
 
+	cbus_endpoint_create(&relay->wal_endpoint,
+			     tt_sprintf("relay_wal_%p", relay),
+			     fiber_schedule_cb, fiber());
+
 	struct relay_is_raft_enabled_msg raft_enabler;
 	if (!relay->replica->anon && relay->version_id >= version_id(2, 6, 0))
 		relay_send_is_raft_enabled(relay, &raft_enabler, true);
@@ -852,7 +862,7 @@ relay_subscribe_f(va_list ap)
 		trigger_add(&relay->r->on_close_log, &on_close_log);
 
 	/* Setup WAL watcher for sending new rows to the replica. */
-	wal_set_watcher(&relay->wal_watcher, relay->endpoint.name,
+	wal_set_watcher(&relay->wal_watcher, relay->wal_endpoint.name,
 			relay_process_wal_event, cbus_process);
 
 	/* Start fiber for receiving replica acks. */
@@ -883,13 +893,15 @@ relay_subscribe_f(va_list ap)
 
 		fiber_cond_wait_deadline(&relay->reader_cond,
 					 relay->last_row_time + timeout);
-
 		/*
 		 * The fiber can be woken by IO cancel, by a timeout of
 		 * status messaging or by an acknowledge to status message.
 		 * Handle cbus messages first.
 		 */
-		cbus_process(&relay->endpoint);
+		inj = errinj(ERRINJ_RELAY_FROM_TX_DELAY, ERRINJ_BOOL);
+		if (inj == NULL || !inj->bparam)
+			cbus_process(&relay->tx_endpoint);
+		cbus_process(&relay->wal_endpoint);
 		/* Check for a heartbeat timeout. */
 		if (ev_monotonic_now(loop()) - relay->last_row_time > timeout)
 			relay_send_heartbeat(relay);
@@ -939,7 +951,8 @@ relay_subscribe_f(va_list ap)
 	/* Destroy cpipe to tx. */
 	cbus_unpair(&relay->tx_pipe, &relay->relay_pipe,
 		    NULL, NULL, cbus_process);
-	cbus_endpoint_destroy(&relay->endpoint, cbus_process);
+	cbus_endpoint_destroy(&relay->wal_endpoint, cbus_process);
+	cbus_endpoint_destroy(&relay->tx_endpoint, cbus_process);
 
 	relay_exit(relay);
 
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index 2c8edfa58..7b71bf375 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -169,6 +169,7 @@ struct errinj {
 	_(ERRINJ_XLOG_META, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_XLOG_READ, ERRINJ_INT, {.iparam = -1}) \
 	_(ERRINJ_XLOG_RENAME_DELAY, ERRINJ_BOOL, {.bparam = false}) \
+	_(ERRINJ_RELAY_FROM_TX_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 
 ENUM0(errinj_id, ERRINJ_LIST);
 extern struct errinj errinjs[];
```
</details>

<details><summary>test</summary>

```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{
    listen = 3313,
    replication = {3313, 3314},
    replication_synchro_quorum = 1,
    replication_timeout = 0.1,
    election_mode = 'manual',
}
box.schema.user.grant('guest', 'super')

-- Step 4
--
box.error.injection.set("ERRINJ_RELAY_FROM_TX_DELAY", true)
box.ctl.promote()
box.info.election.term      -- = 4
box.info.synchro.queue.term -- = 4

-- Step 6
--
box.error.injection.set("ERRINJ_RELAY_FROM_TX_DELAY", false)


--
-- Instance 2
--
-- Step 2
--
box.cfg{
    listen = 3314,
    replication = {3313, 3314},
    read_only = true,
    replication_synchro_quorum = 1,
    replication_timeout = 0.1,
    election_mode = 'manual',
}

-- Step 3
--
box.ctl.promote()

-- Step 5
--
box.info.election.term      -- = 3
box.info.synchro.queue.term -- = 4

-- Step 7
--
box.info.election.term      -- = 4
box.info.synchro.queue.term -- = 4
```
</details>

In the test I do an error injection to simulate too long delivery of TX messages. So as the relay reads WAL messages faster than gets TX messages.

The diff also allows you to implement the proposal about handling the order issue entirely inside of relay. Because now TX and WAL messages are delivered separately. When you meet `PROMOTE` with a too big term, you can spin in `cbus_process(&tx_endpoint)`. Dispatch of `wal_endpoint` won't grab TX messages alongside like it happened when you tried to implement it first time.

It is also worth mentioning that in this test we indeed can observe synchro term > raft term, but I couldn't make any real bug out of it. After you unblock the injection, the raft term is delivered just fine, not ignored.

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:1809]: While you waited for the term broadcast, the term could already be bumped by another node. And below you will write a demote for a wrong term. Same above for promote.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the fixes! Could you please not mark my comments as fixed? I usually look at them before doing a next review to see what was wrong previous time and when all of them are already closed, it complicates the process a bit. I will mark them as resolved myself.

**Comment by grafin** [comment]
Alternative PR #7009

**Comment by sergepetrenko** [comment]
IMO this PR should be closed as superseded by #7009. I'm waiting for @Gerold103's and @grafin's approval before closing it.

**Comment by grafin** [comment]
> this PR should be closed as superseded by https://github.com/tarantool/tarantool/pull/7009.

I agree.

**Pull request closed without merging**
